### PR TITLE
Fix Revert. Бесхмельнова Ксения. Задача 2. Вариант 3. Обедающие философы.

### PR DESCRIPTION
**Pull request by KseniyaBeskhmelnova**
Программа реализует задачу об обедающих философах с использованием MPI для параллельного взаимодействия между процессами. Каждый философ представляет отдельный процесс, который поочередно переходит через состояния THINKING, HUNGRY и EATING. Для перехода в состояние EATING философы запрашивают вилки(ресурсы), отправляя сообщения своим соседям, и освобождают их по завершении. Программа реализует механизм проверки взаимоблокировки (deadlock) и разрешает её, возвращая философа в состояние THINKING. Завершение программы происходит, когда все философы находятся в состоянии THINKING. Такая реализация обеспечивает синхронизацию процессов и управление доступом к разделяемым ресурсам.

Что исправлено:
добавлено internal_order_tests() в pre_processing, validation, run и post_processing

### CODE REVIEW

**Review by Alexey2013 — APPROVED** [review]

**Comment by aobolensk** [comment]
Please, increase code coverage

**Comment by KseniyaBeskhmelnova** [comment]
@aobolensk, done

File: tasks/mpi/beskhmelnova_k_dinning_philosophers/src/dinning_philosophers.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+      break;
+    }
+  }
+  return all_thinking;
+}
+
+template <typename DataType>
+void beskhmelnova_k_dining_philosophers::DiningPhilosophersMPI<DataType>::think() {
+  state = THINKING;
+  std::this_thread::sleep_for(std::chrono::milliseconds(10));
```
Comment (kmichaelk) [inline tasks/mpi/beskhmelnova_k_dinning_philosophers/src/dinning_philosophers.cpp:86]: It seems like these values ​​should be random

File: tasks/mpi/beskhmelnova_k_dinning_philosophers/src/dinning_philosophers.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+template <typename DataType>
+void beskhmelnova_k_dining_philosophers::DiningPhilosophersMPI<DataType>::eat() {
+  state = EATING;
+  std::this_thread::sleep_for(std::chrono::milliseconds(60));
```
Comment (kmichaelk) [inline tasks/mpi/beskhmelnova_k_dinning_philosophers/src/dinning_philosophers.cpp:92]: These and the rest too

File: tasks/mpi/beskhmelnova_k_dinning_philosophers/src/dinning_philosophers.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  state = THINKING;
+  generator.seed(world.rank());
+  distribution = std::uniform_int_distribution<int>(1, 3);
+  return true;
+}
+
+template <typename DataType>
+bool beskhmelnova_k_dining_philosophers::DiningPhilosophersMPI<DataType>::validation() {
+  internal_order_test();
+  int num_philosophers = taskData->inputs_count[0];
```
Comment (kmichaelk) [inline tasks/mpi/beskhmelnova_k_dinning_philosophers/src/dinning_philosophers.cpp:20]: Do you use `num_philosophers` anywhere actually?

**Review by kmichaelk — CHANGES_REQUESTED** [review]
You have a problem in logic somewhere, the execution is being terminated immediately, also the `num_philosophers` you're relying to does not have effect

image

**Comment by kmichaelk** [comment]
Same for perf tests, they don't work and terminate immediately

File: tasks/mpi/beskhmelnova_k_dinning_philosophers/src/dinning_philosophers.cpp [hunk 4]
```diff
@@ -97,15 +96,13 @@ bool beskhmelnova_k_dining_philosophers::DiningPhilosophersMPI<DataType>::check_
...
-  std::this_thread::sleep_for(std::chrono::seconds(think_time));
+  std::this_thread::sleep_for(std::chrono::milliseconds(10));
 }
 
 template <typename DataType>
 void beskhmelnova_k_dining_philosophers::DiningPhilosophersMPI<DataType>::eat() {
   state = EATING;
-  int eat_time = distribution(generator);
-  std::this_thread::sleep_for(std::chrono::seconds(eat_time));
+  std::this_thread::sleep_for(std::chrono::milliseconds(60));
```
Comment (kmichaelk) [inline tasks/mpi/beskhmelnova_k_dinning_philosophers/src/dinning_philosophers.cpp]: ?

**Review by allnes — APPROVED** [review]

**Comment by allnes** [comment]
we will merge after @kmichaelk approve

**Comment by KseniyaBeskhmelnova** [comment]
@kmichaelk, you were right, the processes were terminated before run. i rearranged the completion check and added random in functions. look, please, i hope it is good now.

**Comment by kmichaelk** [comment]
Yes, does not terminate immediately now, but now it behaves incorrectly as all philosophers eat at the same time (which is not possible because everyone needs two forks), this can be confirmed by adding prints and sleep:

```cpp
template <typename DataType>
bool beskhmelnova_k_dining_philosophers::DiningPhilosophersMPI<DataType>::run() {
  internal_order_test();
  while (true) {
    think();
    request_forks();
    printf("[%d] Eat\n", world.rank());
    eat();
    printf(" [%d] Done\n", world.rank());
    std::this_thread::sleep_for(std::chrono::seconds(10));
    release_forks();
    if (check_deadlock()) return false;
    if (check_for_termination()) break;
  }
  printf("[%d] Terminated\n", world.rank());
  return true;
}
```

**Comment by KseniyaBeskhmelnova** [comment]
@kmichaelk, that is what you mean?
Снимок экрана 2024-12-23 221017

**Comment by kmichaelk** [comment]
The current problem with this function is that all processes enter it and exit it all at once (and so proceed with eating, which is not possible by the definition of the dining philosophers task). And this function should block those processes-philosophers who do not have a fork available yet.

In fact, the messages that go in this function should match the messages coming from `release_forks`. It doesn't seem to work as is, maybe try different tags for requesting and releasing forks, you can check if it works by adding the prints and a sleep like in the comment above.

**Comment by kmichaelk** [comment]
The problem seems to be with the tags indeed

File: tasks/mpi/beskhmelnova_k_dinning_philosophers/src/dinning_philosophers.cpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  std::random_device rd;
+  generator.seed(rd() + world.rank());
+  distribution = std::uniform_int_distribution<int>(1, 3);
+  return true;
+}
+
+template <typename DataType>
+bool beskhmelnova_k_dining_philosophers::DiningPhilosophersMPI<DataType>::validation() {
+  internal_order_test();
+  num_philosophers = taskData->inputs_count[0];
```
Comment (kmichaelk) [inline tasks/mpi/beskhmelnova_k_dinning_philosophers/src/dinning_philosophers.cpp:21]: You still don't use it anywhere

**Comment by KseniyaBeskhmelnova** [comment]
please, look again. i added prints like toy and everything is good: 
[7] Eat
 [7] Done
[5] Eat
 [5] Done
[2] Eat
 [2] Done
[4] Eat
 [4] Done
[1] Eat
 [1] Done
[6] Eat
 [6] Done
[3] Eat
 [3] Done
[0] Eat
 [0] Done

**Comment by kmichaelk** [comment]
It seems nothing has changed, at least with the current revision in PR

**Comment by KseniyaBeskhmelnova** [comment]
@kmichaelk, i mean there is something wrong with your prints. everything works correctly on my computer, see the prints above.
You were right about the number of philosophers: in all tests number of philosophers was the same as number of processes. now number of them = number of philosophers in tests.

**Comment by kmichaelk** [comment]
Perhaps you forgot to push the latest commit, currently the problem still persists because you are using the same tag for all messages, everything should be fixed as soon as the request for forks and their return are identified differently, there is very little left to fix.

**Comment by KseniyaBeskhmelnova** [comment]
@kmichaelk, look, please. did i understand you correct?
i hope, it is all right now. it is very hard

**Comment by kmichaelk** [comment]
Interesting task. In fact, you have one synchronization point - it is a pair of `request_forks` and `release_forks` - `request_forks` should block the process until a message from `release_forks` arrives there. Because the memory is not shared, there is a chicken and egg problem - at the very beginning all processes should call `release_forks`, but then each process will receive a message and they will not block in `request_forks`, **which will lead to all philosophers eating at the same time**, which is wrong. 

That is, the problem comes down to the fact that in the current implementation, a fork is not a shared resource, so it can be "in the hands" of two philosopher processes at once, which is impossible by definition.

I just thought, IMO, this can only be properly resolved by introducing a separate coordinator ("table") process, or by dedicating a separate process per each fork.

ALTHOUGH there is another possible partial solution, it seems to me (de facto inlined the `release_forks` function into `run`, right forks are dropped on the beginning only to the part of processes):
image
image

And it gives correct output now:

```
[1] Eat()
[7] Eat()
[5] Eat()
[3] Eat()
 [1] Done
[2] Eat()
 [7] Done
[0] Eat()
 [5] Done
[6] Eat()
 [3] Done
[4] Eat()
 [0] Done
 [2] Done
 [4] Done
 [6] Done
[1] Eat()
[3] Eat()
[5] Eat()
[7] Eat()
 [1] Done
[2] Eat()
 [7] Done
 [3] Done
 [5] Done
[6] Eat()
[0] Eat()
[4] Eat()
 [2] Done
 [0] Done
 [6] Done
 [4] Done
[5] Eat()
[3] Eat()
[1] Eat()
[7] Eat()
 [5] Done
[6] Eat()
 [3] Done
[4] Eat()
 [1] Done
[2] Eat()
[0] Eat()
 [7] Done
 [6] Done
 [4] Done
 [2] Done
 [0] Done
[1] Eat()
[7] Eat()
[5] Eat()
[3] Eat()
 [1] Done
[2] Eat()
 [7] Done
 [5] Done
[6] Eat()
[0] Eat()
 [3] Done
[4] Eat()
 [0] Done
 [2] Done
 [6] Done
 [4] Done
[3] Eat()
[1] Eat()
[5] Eat()
[7] Eat()
```

And for infinity as there's no exit condition as for now, but works properly judging by the ranks (two neighboring philosophers do not eat at the same time).

If the current revision is considered acceptable, I would approve, but otherwise I would abstain from voting in either direction.

**Review by kmichaelk — APPROVED** [review]
The workaround I suggested above is also flawed, because it will never lead to combinations like this (green philosophers are eating). In fact, the same sets of philosophers will switch each time, so a coordinator process is needed for correct solution.

image

Not a fan of unwittingly force someone to redo a hard task at the end but JFYI it doesn't work correctly if you're interested

**Comment by KseniyaBeskhmelnova** [comment]
@aobolensk, @allnes, you can merge now.

it was very hard and took a lot of time, but i did :)

**Comment by KseniyaBeskhmelnova** [comment]
if I fail the exam today, it will be partly on Mikhail's conscience )))

**Pull request merged**